supplementary/alica_ros1/alica_ros_turtlesim/test/test_main.cpp

```diff
@@ -0,0 +1,183 @@
+#include <std_msgs/Empty.h>
+#include <turtlesim/Pose.h>
+
+#include <autodiff/AutoDiff.h>
+#include <gtest/gtest.h>
+#include <ros/ros.h>
+
+#include <chrono>
+#include <thread>
+
+class TurtlePosition;
+
+#define RUN_UNTIL_ASSERT_EQ(left, right, ms)                                                                                                                   \
+    do {                                                                                                                                                       \
+        using namespace std::chrono;                                                                                                                           \
+        auto start = system_clock::now();                                                                                                                      \
+        do {                                                                                                                                                   \
+            ros::spinOnce();                                                                                                                                   \
+            if (left == right) {                                                                                                                               \
+                ASSERT_EQ(left, right);                                                                                                                        \
```

Comment:
can't be false (unless it's atomic and other thread could change between this line and previous one)

Reply:
This is a macro, left and right will be reevaluated each spin

Reply:
@ntfshard you're correct, I was assuming 
```cpp
if (left == right) {                                                                                                                               \
    ASSERT_EQ(left, right);                                                                                                              \
```
is an atomic operation, which for what is happening in this test, it is. I'm using left and right as aliases for things that are either constants or something that will be changing values from one ros spin to another, so my assumption is correct for this case. 

In theory, we can have left and right expressions returning values different from the previous line because those can be expressions getting values from some object that is being modified by another thread since the macro is for generic situations. 

My goal with this is just to check if a value is equal at some moment and then make sure it will be asserted (here I want to know if the number of subscribers is 1 at some point, can't check for the validity of this during the entire test run unless I assert this every time until the end of the program (which is bad)). So because of this, I propose this

```cpp
#define RUN_UNTIL_ASSERT_EQ(left, right, ms)                                                                                                                   \
    do {                                                                                                                                                       \
        using namespace std::chrono;                                                                                                                           \
        auto start = system_clock::now();                                                                                                                      \
        do {                                                                                                                                                   \
            ros::spinOnce();                                                                                                                                   \
            if (left == right) {                                                                                                                               \
                break;                                                                                                                                         \
            }                                                                                                                                                  \
            ros::Duration(0.01).sleep();                                                                                                                       \
        } while (duration_cast<milliseconds>(system_clock::now() - start).count() < ms);                                                                       \
        ASSERT_EQ(left, right);                                                                                                                                \
    } while (0)
```
Once `(left == right)` I don't have to assert anything, the condition I wanted was met once and I will not assert anything true (I could put ASSERT_TRUE(true), but there's no point). However, I will assert when I have a timeout because it can be false at that moment.

What do you think?

Reply:
I think this is a nit, please address in another PR if you have to

Reply:
I see the issue, yeah the change makes sense but I don't think it matters much